File: 2023/day19/solution.py
#!/usr/bin/env python3
# type: ignore
# FIXME: dump the tuples, reenable mypy

import sys
import re

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True

# yapf: disable
for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i': FILENAME = 'input.txt'
      case '-p1': PART1, PART2 = True, False
      case '-p2': PART1, PART2 = False, True
      case '-p0': PART1, PART2 = False, False
      case _: raise Exception(f'unknown {arg=}')

WORKFLOWS_RE = re.compile(r'^([a-z]+)\{(.+)\}$', re.MULTILINE)
RULES_RE = re.compile(r'(?:([xmas])([<>])([0-9+]+):)?(R|A|[a-z]+),?')
PARTS_RE = re.compile(r'^\{x=([0-9]+),m=([0-9]+),a=([0-9]+),s=([0-9]+)\}$', re.MULTILINE)
with open(FILENAME, 'r') as f:
  sections = f.read().split('\n\n')
workflows = {
  w.group(1): [
    (
      (
        {'x': 0, 'm': 1, 'a': 2, 's': 3}[r.group(1)],
        r.group(2),
        int(r.group(3)),
      ) if r.group(1) else (),
      r.group(4),
    )
    for r in re.finditer(RULES_RE, w.group(2))
  ]
  for w in re.finditer(WORKFLOWS_RE, sections[0])
}
#yapf: enable
# tuple so it can be memoised
parts = [
  (
    int(p.group(1)),
    int(p.group(2)),
    int(p.group(3)),
    int(p.group(4)),
  ) for p in re.finditer(PARTS_RE, sections[1])
]

# if DEBUG > 1:
#   print(f'{workflows=}')
#   print(f'{parts=}')


def solve(part: tuple[int, int, int, int], workflow_name: str = 'in') -> bool:
  workflow = workflows[workflow_name]
  for rule in workflow:
    valid = True
    if rule[0]:
      prop = part[rule[0][0]]
      op = rule[0][1]
      val = rule[0][2]
      valid = (op == '>' and prop > val) or (op == '<' and prop < val)
    if valid:
      res = rule[1]
      match res:
        case 'A':
          return True
        case 'R':
          return False
      return solve(part, res)
  # mypy pipe down pls
  assert False


def solve3(subset: list[tuple[int, int]], workflow_name: str = 'in') -> int:
  if DEBUG > 0: print(f'{subset=} {workflow_name=}')
  if workflow_name == 'A':
    # hopefully the count of possible subset entries
    size = 1
    for part in subset:
      # off by one clown strikes again
      size *= part[1] - part[0] + 1
    if DEBUG > 1: print(f'  accept {size=}')
    return size
  if workflow_name == 'R': return 0
  workflow, count = workflows[workflow_name], 0
  for rule in workflow:
    if DEBUG > 1: print(f'  {rule=}')
    dest = rule[1]
    if rule[0]:
      part_ix = rule[0][0]
      part_from = subset[part_ix][0]
      part_to = subset[part_ix][1]
      op = rule[0][1]
      val = rule[0][2]
      #FIXME: might validate that val+-1 is not out of range for partial matches
      if op == '<' and part_from < val:
        # complete match - we can return here
        if part_to < val: return count + solve3(subset, dest)
        else:
          # partial match. pass matched part to dest, remainder to next rule
          match_subset = [*subset]
          match_subset[part_ix] = (part_from, val - 1)
          count += solve3(match_subset, dest)
          subset[part_ix] = (val, part_to)
      elif op == '>' and part_to > val:
        # complete match - we can return here
        if part_from > val: return count + solve3(subset, dest)
        else:
          # partial match. pass matched part to dest, remainder to next rule
          match_subset = [*subset]
          match_subset[part_ix] = (val + 1, part_to)
          count += solve3(match_subset, dest)
          subset[part_ix] = (part_from, val)
      # else next rule
    else:
      # conditionless is the last rule so we can return here
      return count + solve3(subset, dest)
  assert False


def part1():
  total = 0
  for part in parts:
    if solve(part):
      total += part[0] + part[1] + part[2] + part[3]
  print(f'part 1: {total=}')


def part2():
  result = solve3([(1, 4000)] * 4)
  print(f'part 2: {result}')
  # 123331556462603

if PART1: part1()
if PART2: part2()

File: 2023/day19/solution.ts
#!/usr/bin/env -S deno --allow-read
// #region base aoc template
class Args {
  constructor(public filename = 'input.txt', public debug = 0, public part1 = true, public part2 = true) {
    for (const arg of Deno.args) {
      const [key, value] = [arg.slice(0, 2), (fallback = 0) => Number(arg.slice(2) || fallback)];
      if (key === '-d') debug = value(1);
      else if (key === '-e') filename = `example${arg.slice(2)}.txt`;
      else if (key === '-i') filename = 'input.txt';
      else if (key === '-p') [part1, part2] = [(value(0) & 1) === 1, (value(0) & 2) === 2];
      else throw new Error(`unrecognised arg="${arg}"`);
    }
    [this.filename, this.debug, this.part1, this.part2] = [filename, debug, part1, part2];
    console.log(`args: {filename: "${filename}", debug: ${debug}, part1: ${part1}, part2: ${part2} }`);
  }
}

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (args.debug >= level) console.debug(...data);
};

const args = new Args();
const input = await Deno.readTextFile(args.filename);
// #endregion

type PartProperty = 'x' | 'm' | 'a' | 's';
type RuleComparator = '<' | '>';
interface Rule {
  property: PartProperty;
  comparator: RuleComparator;
  value: number;
  nextWorkflow: string;
}
interface Workflow {
  rules: Rule[];
  fallback: string;
}
interface Part {
  x: number;
  m: number;
  a: number;
  s: number;
}
interface Range {
  from: number;
  to: number;
}
interface PartRange {
  x: Range;
  m: Range;
  a: Range;
  s: Range;
}

const parseInput = () => {
  const [workflowsSection, partsSection] = input.split('\n\n');

  const workflowRegex = /^([a-z]+)\{(.*),(R|A|[a-z]+)\}$/gm;
  const ruleRegex = /([xmas])([<>])(\d+):(R|A|[a-z]+)(?:,|$)/g;

  const workflows = new Map<string, Workflow>(
    workflowsSection.matchAll(workflowRegex).map((workflowMatch) => [
      workflowMatch[1],
      {
        rules: workflowMatch[2]
          .matchAll(ruleRegex)
          .map((ruleMatch) => ({
            property: ruleMatch[1] as PartProperty,
            comparator: ruleMatch[2] as RuleComparator,
            value: Number(ruleMatch[3]),
            nextWorkflow: ruleMatch[4],
          }))
          .toArray(),
        fallback: workflowMatch[3],
      },
    ])
  );
  debug(3, workflows);

  const partRegex = /^\{x=(\d+),m=(\d+),a=(\d+),s=(\d+)\}$/gm;

  const parts: Part[] = partsSection
    .matchAll(partRegex)
    .map((partMatch) => ({
      x: Number(partMatch[1]),
      m: Number(partMatch[2]),
      a: Number(partMatch[3]),
      s: Number(partMatch[4]),
    }))
    .toArray();
  debug(3, parts);

  return { parts, workflows };
};

const testPart = (part: Part, workflows: Map<string, Workflow>, workflowName: string = 'in') => {
  debug(1, { part, workflowName });
  if (workflowName === 'R') return false;
  if (workflowName === 'A') return true;
  const workflow = workflows.get(workflowName)!;
  for (const rule of workflow.rules) {
    if (
      (rule.comparator === '<' && part[rule.property] < rule.value) ||
      (rule.comparator === '>' && part[rule.property] > rule.value)
    )
      return testPart(part, workflows, rule.nextWorkflow);
  }
  return testPart(part, workflows, workflow.fallback);
};

const testRange = (range: PartRange, workflows: Map<string, Workflow>, workflowName: string = 'in'): number => {
  debug(1, { range, workflowName });
  if (workflowName === 'R') return 0;
  if (workflowName === 'A')
    return (
      (range.x.to - range.x.from + 1) *
      (range.m.to - range.m.from + 1) *
      (range.a.to - range.a.from + 1) *
      (range.s.to - range.s.from + 1)
    );
  const workflow = workflows.get(workflowName)!;
  let count = 0;
  for (const rule of workflow.rules) {
    // count+=testRange(matchedRange), range=unmatchedRange
    if (rule.comparator === '<' && range[rule.property].from < rule.value) {
      if (range[rule.property].to < rule.value) return count + testRange(range, workflows, rule.nextWorkflow);
      count += testRange(
        { ...range, [rule.property]: { ...range[rule.property], to: rule.value - 1 } },
        workflows,
        rule.nextWorkflow
      );
      range = { ...range, [rule.property]: { ...range[rule.property], from: rule.value } };
    }
    if (rule.comparator === '>' && range[rule.property].to > rule.value) {
      if (range[rule.property].from > rule.value) return count + testRange(range, workflows, rule.nextWorkflow);
      count += testRange(
        { ...range, [rule.property]: { ...range[rule.property], from: rule.value + 1 } },
        workflows,
        rule.nextWorkflow
      );
      range = { ...range, [rule.property]: { ...range[rule.property], to: rule.value } };
    }
  }
  return count + testRange(range, workflows, workflow.fallback);
};

const part1 = () => {
  const { parts, workflows } = parseInput();
  const total = parts.reduce((acc, part) => (testPart(part, workflows) ? acc + part.x + part.m + part.a + part.s : acc), 0);
  console.log('part 1:', total);
  // 418498
};

const part2 = () => {
  const { workflows } = parseInput();
  const total = testRange(
    { x: { from: 1, to: 4000 }, m: { from: 1, to: 4000 }, a: { from: 1, to: 4000 }, s: { from: 1, to: 4000 } },
    workflows
  );
  console.log('part 2:', total);
  // 123331556462603
};

if (args.part1) part1();
if (args.part2) part2();
